### PR TITLE
Deprecation warning on CommonChunksPlugin. Add SplitChunksPlugin page

### PR DESCRIPTION
**Pull request by jeremenichelli**
Adding Tobias as a reviewer since I used his gist as a base for the SplitChunksPlugin page. I've also added a deprecation warning on CommonChunksPlugin one (not deleting it for now since developers and teams who can immediately migrate might need it).

### CODE REVIEW

**Comment by montogeek** [comment]
Great initial work!

**Comment by jeremenichelli** [comment]
@skipjack I want to bother you as less as possible, but tagging you here since it's a whole new page that would love to have a look from you.

File: src/content/plugins/commons-chunk-plugin.md [hunk 1]
```diff
@@ -6,9 +6,14 @@ contributors:
   - christopher4lis
   - kevinzwhuang
   - jdbevan
+  - jeremenichelli
 ---
 
-The `CommonsChunkPlugin` is an opt-in feature that creates a separate file (known as a chunk), consisting of common modules shared between multiple entry points. By separating common modules from bundles, the resulting chunked file can be loaded once initially, and stored in cache for later use. This results in pagespeed optimizations as the browser can quickly serve the shared code from cache, rather than being forced to load a larger bundle whenever a new page is visited.
+The `CommonsChunkPlugin` is an opt-in feature that creates a separate file (known as a chunk), consisting of common modules shared between multiple entry points.
+
+W> CommonsChunkPlugin has been deprecated in webpack v4 legato. If you want to know how chunks are treated in the newest version checkout [SplitChunksPlugin](/plugins/split-chunks-plugin/).
```
Comment (skipjack) [inline src/content/plugins/commons-chunk-plugin.md]: I would rephrase this slightly:

> The `CommonsChunkPlugin` has been deprecated in webpack v4 legato.. To learn how chunks are treated in the latest version, check out the [`SplitChunksPlugin`](/plugins/split-chunks-plugin).

File: src/content/plugins/split-chunks-plugin.md [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+  - sokra
+  - jeremenichelli
+related:
+  - title: "webpack 4: Code Splitting, chunk graph and the splitChunks optimization"
+    url: https://medium.com/webpack/webpack-4-code-splitting-chunk-graph-and-the-splitchunks-optimization-be739a861366
+---
+
+Originally in webpack, chunks (and modules import inside them) were connected by a parent-child relationship in the internal webpack graph.
+
+This connection didn't allow to further optimizations and reluted into more coded downloaded.
```
Comment (skipjack) [inline src/content/plugins/split-chunks-plugin.md]: s/reluted/resulted

Comment (skipjack) [reply src/content/plugins/split-chunks-plugin.md]: s/more coded downloaded/more downloaded code

File: src/content/plugins/split-chunks-plugin.md [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+	chunks: "async",
+	minSize: 30000,
+	minChunks: 1,
+	maxAsyncRequests: 5,
+	maxInitialRequests: 3,
+	name: true,
+	cacheGroups: {
+		default: {
+			minChunks: 2,
+			priority: -20
```
Comment (mike1808) [inline src/content/plugins/split-chunks-plugin.md]: comma is missing

File: src/content/plugins/split-chunks-plugin.md [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+
+### Example 2
+
+Create a `vendors` chunk, which includes all code from node_modules in the whole application.
+
+``` js
+splitChunks: {
+	cacheGroups: {
+		commons: {
+			test: /[\\/]node_modules[\\/]
```
Comment (robert-claypool) [inline src/content/plugins/split-chunks-plugin.md]: Missing a slash-comma: `test: /[\\/]node_modules[\\/]/,`

File: src/content/plugins/split-chunks-plugin.md [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+
+This connection didn't allow to further optimizations and reluted into more coded downloaded.
+
+webpack v4 removes the `CommonsChunkPlugin` in favor of `optimization.splitChunks` and `optimization.runtimeChunk` options. Here is how the new flow works.
+
+## Defaults
+
+Out of the box `SplitChunksPlugin` should work great for most users.
+
+By default the plugin only affect on-demand chunks, because changing initial chunks would affect the script tags that the HTML file should include to run the project.
```
Comment (blowery) [inline src/content/plugins/split-chunks-plugin.md]: `affect` should be `affects`

File: src/content/plugins/split-chunks-plugin.md [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+
+Putting the content of `helpers` into each chunk will result into its code being downloaded twice. By using a separate chunk this will only happen once. We pay the cost of an additional request, which could be considered a tradeoff. That's why there is a minimum size of 30kb.
+
+T> With the new `optimizations.splitChunks.chunks: "all"` option the same would happend for initial chunks. Chunks can even be shared between entrypoints and on-demand loading.
+
+## Configuration
+
+For these people that like to have more control over this functionality, webpack provides a set of options to better fit your needs.
+
+W> If you are manually changing the split configuration, measure the impact of the changes to see and make sure there's a real benefit. The defaults are choosen to fit best practices of web performance.
```
Comment (benthemonkey) [inline src/content/plugins/split-chunks-plugin.md]: s/choosen/chosen
(change below my opinion, not related to grammar):
s/fit best practices of web performance/fit web performance best practices

File: src/content/plugins/split-chunks-plugin.md [hunk 7]
```diff
@@ -0,0 +1,220 @@
...
+
+### Configuring cache groups
+
+The defaults assigns all modules from `node_modules` to a cache group called `vendors` and all modules duplicated in at least 2 chunks to a change group `default`.
```
Comment (benthemonkey) [inline src/content/plugins/split-chunks-plugin.md]: s/change group/cache group

**Comment by benthemonkey** [comment]
I notice these docs don't include any information about the `enforce` cache group setting (undocumented in Sokra's gist). I've been playing around with that option and was hoping to learn more about it from these docs. It seems like it is useful for overriding the base settings for only one cache group. E.g.:

```
splitChunks: {
	chunks: "async",
	minSize: 30000,
	minChunks: 1,
	maxAsyncRequests: 5,
	maxInitialRequests: 3,
	name: true,
	cacheGroups: {
		default: {
			minChunks: 2,
			priority: -20
			reuseExistingChunk: true,
		},
		vendors: {
			test: /[\\/]node_modules[\\/]/,
			priority: -10
		},
		special: {
			test: myTestFunc,
			enforce: true
		}
	}
}
```

Please correct me if I am mistaken: The `enforce` allows the `special` cache group to ignore the `minSize`, `minChunks`, `maxAsyncRequests` and `maxInitialRequests` set in the overall cacheGroups config and purely use `test` to decide what goes into the chunk. What I'm unclear on is how that fits in when used with `reuseExistingChunk`.

File: src/content/plugins/split-chunks-plugin.md [hunk 8]
```diff
@@ -0,0 +1,220 @@
...
+		}
+	}
+}
+```
+
+W> This downloads more code than neccessary.
+
+## `optimization.runtimeChunk`
+
+Setting `optimization.runtimeChunk` to `true` adds an additonal chunk to each entrypoint containing only the runtime.
```
Comment (benthemonkey) [inline src/content/plugins/split-chunks-plugin.md]: include the different things that you can pass to runtimeChunk. It can be a boolean, `'single'`, `'multiple'` or and object that has property `name` whose value can be either a string or a function of signature `(entrypoint: Chunk) => string`

File: src/content/plugins/split-chunks-plugin.md [hunk 9]
```diff
@@ -0,0 +1,220 @@
...
+By default the plugin only affect on-demand chunks, because changing initial chunks would affect the script tags that the HTML file should include to run the project.
+
+webpack will automatically split chunks based on these conditions:
+
+* New chunk can be shared OR modules are from the `node_modules` folder
+* New chunk would be bigger than 30kb (before min+gz)
+* Maximum number of parallel request when loading chunks on demand would be lower or equal to 5
+* Maximum number of parallel request at initial page load would be lower or equal to 3
+
+When trying to fullfill the last two conditions, bigger chunks are preferred.
```
Comment (balanceiskey) [inline src/content/plugins/split-chunks-plugin.md]: `fullfill` -> `fulfill`

File: src/content/plugins/commons-chunk-plugin.md (hunk 1 above)
Comment (audiolion) [reply src/content/plugins/commons-chunk-plugin.md]: s/legato../legato.

File: src/content/plugins/split-chunks-plugin.md [hunk 10]
```diff
@@ -0,0 +1,220 @@
...
+## Defaults
+
+Out of the box `SplitChunksPlugin` should work great for most users.
+
+By default the plugin only affect on-demand chunks, because changing initial chunks would affect the script tags that the HTML file should include to run the project.
+
+webpack will automatically split chunks based on these conditions:
+
+* New chunk can be shared OR modules are from the `node_modules` folder
+* New chunk would be bigger than 30kb (before min+gz)
```
Comment (lencioni) [inline src/content/plugins/split-chunks-plugin.md]: kb -> KiB (here and elsewhere)

File: src/content/plugins/split-chunks-plugin.md (hunk 5 above)
Comment (audiolion) [reply src/content/plugins/split-chunks-plugin.md]: @blowery I don't agree, affect is correct grammar

File: src/content/plugins/split-chunks-plugin.md [hunk 11]
```diff
@@ -0,0 +1,220 @@
...
+* `maxInitialRequests` (default 3) Maximum number of parallel requests at an entrypoint
+* `maxAsyncRequests` (default 5) Maximum number of parallel requests at on-demand loading
+
+### Naming
+
+To control the chunk name of the split chunk the `name` option can be used.
+
+W> When assigning equal names to different split chunks they are merged together. This can be used to put all vendor modules into a single shared chunk, but it's not recommend since it can result in more code downloaded.
+
+The magic value `true` automatically chooses a name based on chunks and cache group key. Elsewise a string or function can be passed.
```
Comment (audiolion) [inline src/content/plugins/split-chunks-plugin.md]: s/Elsewise/Otherwise

[Elsewise usage in english](https://www.collinsdictionary.com/us/dictionary/english/elsewise)

File: src/content/plugins/split-chunks-plugin.md (hunk 8 above)
Comment (jeremenichelli) [reply src/content/plugins/split-chunks-plugin.md]: Hey @sokra can you fill in some technical gap here? Should we add `enforce` details in this piece, I would need your help with it 😓

**Comment by jeremenichelli** [comment]
Thanks everyone for all your feedback. Working on updates and getting changes pushed soon.

File: src/content/plugins/commons-chunk-plugin.md [hunk 12]
```diff
@@ -6,9 +6,14 @@ contributors:
...
+
+By separating common modules from bundles, the resulting chunked file can be loaded once initially, and stored in cache for later use. This results in pagespeed optimizations as the browser can quickly serve the shared code from cache, rather than being forced to load a larger bundle whenever a new page is visited.
 
 ```javascript
 new webpack.optimize.CommonsChunkPlugin(options)
```
Comment (audiolion) [inline src/content/plugins/commons-chunk-plugin.md]: shouldn't this be removed? or replaced with:
```json
optimization: {
  splitChunks: {
    name: 'all',
  },
}
```

Comment (jeremenichelli) [reply src/content/plugins/commons-chunk-plugin.md]: This is the old CommonsChunkPlugin page. We are keeping the old documentaiton for teams that are already using it and haven't migrated to v4. We are adding a warning on top talking about the deprecation of it from now on.

File: src/content/plugins/split-chunks-plugin.md [hunk 13]
```diff
@@ -0,0 +1,220 @@
...
+		}
+	}
+}
+```
+
+By default cache groups inherit options from `splitChunks.*`, but `test`, `priority` and `reuseExistingChunk` can only be configured on cache group level.
+
+`cacheGroups` is an object where keys are the cache group names. All options from the ones listed above are possible: `chunks`, `minSize`, `minChunks`, `maxAsyncRequests`, `maxInitialRequests`, `name`.
+
+You can set `optimization.splitChunks.cacheGroups.default` to `false` to disable the default cache group.
```
Comment (sokra) [inline src/content/plugins/split-chunks-plugin.md]: Same for `optimization.splitChunks.cacheGroups.vendors`.

File: src/content/plugins/split-chunks-plugin.md [hunk 14]
```diff
@@ -0,0 +1,220 @@
...
+
+The priority of the default groups are negative to allow any custom cache group to take higher priority (the default value is `0`).
+
+Here are some examples and their effect:
+
+### Example 1
```
Comment (sokra) [inline src/content/plugins/split-chunks-plugin.md]: Add another example:

Enable the default optimizations for all chunks instead of only async chunks. This is the recommended configuration.

``` js
splitChunks: {
  chunks: "all"
}
```

File: src/content/plugins/split-chunks-plugin.md [hunk 15]
```diff
@@ -0,0 +1,220 @@
...
+splitChunks: {
+	cacheGroups: {
+		commons: {
+			name: "commons",
+			chunks: "initial",
+			minChunks: 2
+		}
+	}
+}
+```
```
Comment (sokra) [inline src/content/plugins/split-chunks-plugin.md]: By default this configuration will only create a commons chunk when the minimum size and maximum number of requests are valid. Adding `enforce: true`, will disable any not configured condition and enforce creating the chunk.

File: src/content/plugins/split-chunks-plugin.md [hunk 16]
```diff
@@ -0,0 +1,220 @@
...
+
+By default cache groups inherit options from `splitChunks.*`, but `test`, `priority` and `reuseExistingChunk` can only be configured on cache group level.
+
+`cacheGroups` is an object where keys are the cache group names. All options from the ones listed above are possible: `chunks`, `minSize`, `minChunks`, `maxAsyncRequests`, `maxInitialRequests`, `name`.
+
+You can set `optimization.splitChunks.cacheGroups.default` to `false` to disable the default cache group.
+
+The priority of the default groups are negative to allow any custom cache group to take higher priority (the default value is `0`).
```
Comment (sokra) [inline src/content/plugins/split-chunks-plugin.md]: The `priority` controls the order in which cache groups are executed. Higher priorities will execute before lower priorities.

File: src/content/plugins/split-chunks-plugin.md [hunk 17]
```diff
@@ -0,0 +1,220 @@
...
+title: SplitChunksPlugin
+contributors:
+  - sokra
+  - jeremenichelli
+related:
+  - title: "webpack 4: Code Splitting, chunk graph and the splitChunks optimization"
+    url: https://medium.com/webpack/webpack-4-code-splitting-chunk-graph-and-the-splitchunks-optimization-be739a861366
+---
+
+Originally in webpack, chunks (and modules import inside them) were connected by a parent-child relationship in the internal webpack graph.
```
Comment (edmorley) [inline src/content/plugins/split-chunks-plugin.md]: s/modules import/modules imported/

File: src/content/plugins/split-chunks-plugin.md (hunk 2 above)
Comment (edmorley) [reply src/content/plugins/split-chunks-plugin.md]: s/allow to/allow for/

File: src/content/plugins/split-chunks-plugin.md [hunk 18]
```diff
@@ -0,0 +1,220 @@
...
+
+This connection didn't allow to further optimizations and reluted into more coded downloaded.
+
+webpack v4 removes the `CommonsChunkPlugin` in favor of `optimization.splitChunks` and `optimization.runtimeChunk` options. Here is how the new flow works.
```
Comment (edmorley) [inline src/content/plugins/split-chunks-plugin.md]: s/favor of/favor of the/

File: src/content/plugins/split-chunks-plugin.md (hunk 5 above)
Comment (edmorley) [reply src/content/plugins/split-chunks-plugin.md]: @audiolion, are you sure? I think if "plugin" were plural, then I agree "affect" should be used (ie "the plugins only affect on-demand chunks"), however given it's singular, "affects" seems more correct to me.

File: src/content/plugins/split-chunks-plugin.md [hunk 19]
```diff
@@ -0,0 +1,220 @@
...
+Out of the box `SplitChunksPlugin` should work great for most users.
+
+By default the plugin only affect on-demand chunks, because changing initial chunks would affect the script tags that the HTML file should include to run the project.
+
+webpack will automatically split chunks based on these conditions:
+
+* New chunk can be shared OR modules are from the `node_modules` folder
+* New chunk would be bigger than 30kb (before min+gz)
+* Maximum number of parallel request when loading chunks on demand would be lower or equal to 5
+* Maximum number of parallel request at initial page load would be lower or equal to 3
```
Comment (edmorley) [inline src/content/plugins/split-chunks-plugin.md]: s/parallel request/parallel requests/ (x2)

File: src/content/plugins/split-chunks-plugin.md [hunk 20]
```diff
@@ -0,0 +1,220 @@
...
+* Condition 3: Number of parallel requests at the import calls is 2
+* Condition 4: Doesn't affect request at initial page load
+
+Putting the content of `helpers` into each chunk will result into its code being downloaded twice. By using a separate chunk this will only happen once. We pay the cost of an additional request, which could be considered a tradeoff. That's why there is a minimum size of 30kb.
+
+T> With the new `optimizations.splitChunks.chunks: "all"` option the same would happend for initial chunks. Chunks can even be shared between entrypoints and on-demand loading.
+
+## Configuration
+
+For these people that like to have more control over this functionality, webpack provides a set of options to better fit your needs.
```
Comment (edmorley) [inline src/content/plugins/split-chunks-plugin.md]: s/For these people/For people/

File: src/content/plugins/split-chunks-plugin.md (hunk 5 above)
Comment (audiolion) [reply src/content/plugins/split-chunks-plugin.md]: haha, I was reading the wrong "affect", `initial chunks would affect(s) the script tags...`

File: src/content/plugins/split-chunks-plugin.md [hunk 21]
```diff
@@ -0,0 +1,220 @@
...
+
+With the `chunks` option the selected chunks can be configured. There are 3 values possible `"initial"`, `"async"` and `"all"`. When configured the optimization only selects initial chunks, on-demand chunks or all chunks.
+
+The option `reuseExistingChunk` allows to reuse existing chunks instead of creating a new one when modules match exactly.
+
+This can be controlled per cache group.
+
+### Select modules
+
+The `test` option controls which modules are selected by this cache group. Omitting it selects all modules. It can be a RegExp, string or function.
```
Comment (lencioni) [inline src/content/plugins/split-chunks-plugin.md]: It would be nice if this section went more into details about the function version of this option, e.g. what the args to the function are, how to use them, and what the function should return.

File: src/content/plugins/split-chunks-plugin.md (hunk 5 above)
Comment (jeremenichelli) [reply src/content/plugins/split-chunks-plugin.md]: Lol, same happened to me. Thanks.

File: src/content/plugins/split-chunks-plugin.md [hunk 22]
```diff
@@ -0,0 +1,220 @@
...
+			test: /[\\/]node_modules[\\/]/,
+			priority: -10
+		}
+	}
+}
+```
+
+By default cache groups inherit options from `splitChunks.*`, but `test`, `priority` and `reuseExistingChunk` can only be configured on cache group level.
+
+`cacheGroups` is an object where keys are the cache group names. All options from the ones listed above are possible: `chunks`, `minSize`, `minChunks`, `maxAsyncRequests`, `maxInitialRequests`, `name`.
```
Comment (lencioni) [inline src/content/plugins/split-chunks-plugin.md]: It looks like `cacheGroups` can also be a function? Should this be documented here? https://github.com/webpack/webpack/blob/66ff412f7e2409ee535c2dc79abf4068015b2e22/lib/optimize/SplitChunksPlugin.js#L132

File: src/content/plugins/split-chunks-plugin.md (hunk 21 above)
Comment (lencioni) [reply src/content/plugins/split-chunks-plugin.md]: FWIW, it looks like `test` fn is called with 2 args. `module` and `chunks`. 

https://github.com/webpack/webpack/blob/66ff412f7e2409ee535c2dc79abf4068015b2e22/lib/optimize/SplitChunksPlugin.js#L188

I think `module` is a string that is the actual code in the module. `chunks` seems to be an array  of objects with metadata about each chunk. I'm not sure where this object is documented or defined in code yet.

It seems that `test` functions should return a boolean.

Comment (lencioni) [reply src/content/plugins/split-chunks-plugin.md]: Ah found it https://github.com/webpack/webpack/blob/fde018300aa52262c384e937c408d5dd97d62951/lib/Chunk.js#L46

Comment (jeremenichelli) [reply src/content/plugins/split-chunks-plugin.md]: Yes, I imagine that boolean decides if the current module goes to a chunk or not.

File: src/content/plugins/split-chunks-plugin.md (hunk 8 above)
Comment (marcofugaro) [reply src/content/plugins/split-chunks-plugin.md]: What is its default value? In this [blog post](https://medium.com/webpack/webpack-4-mode-and-optimization-5423a6bc597a) it says it's `true`, but from trying it out it's not.

File: src/content/plugins/split-chunks-plugin.md [hunk 23]
```diff
@@ -0,0 +1,220 @@
...
+Why:
+
+* Condition 1: The chunk is shared between both import calls
+* Condition 2: `helpers` is bigger than 30kb
+* Condition 3: Number of parallel requests at the import calls is 2
+* Condition 4: Doesn't affect request at initial page load
+
+Putting the content of `helpers` into each chunk will result into its code being downloaded twice. By using a separate chunk this will only happen once. We pay the cost of an additional request, which could be considered a tradeoff. That's why there is a minimum size of 30kb.
+
+T> With the new `optimizations.splitChunks.chunks: "all"` option the same would happend for initial chunks. Chunks can even be shared between entrypoints and on-demand loading.
```
Comment (ind1go) [inline src/content/plugins/split-chunks-plugin.md]: s/optimizations/optimization/

File: src/content/plugins/split-chunks-plugin.md (hunk 14 above)
Comment (jeremenichelli) [reply src/content/plugins/split-chunks-plugin.md]: I've added a specific section for `chunks: "all"` so we highlight this configuration set as recommended, pushing in a bit.

**Comment by jeremenichelli** [comment]
Thanks everyone for the comments, I'm creating a new pull request since the `next` branch has been merged and rebase is completely broken from here in files that haven't been touched and I don't want to cause regression on content.

@sokra I will create an issue with some stuff that needs to be revisited or cover in the split chunks article. Thanks for the comments.
